flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsCoordinatorContext.java

```diff
@@ -0,0 +1,172 @@
+          int parallelism = currentParallelism();
+          for (int i = 0; i < parallelism; ++i) {
+            subtaskGateways.getOnlyGatewayAndCheckReady(i).sendEvent(dataStatisticsEvent);
+          }
+          return null;
+        },
+        String.format(
+            "Failed to send global data statistics %s for checkpoint %d",
+            globalDataStatistics, checkpointId));
+  }
+
+  int currentParallelism() {
+    return operatorCoordinatorContext.currentParallelism();
+  }
+
+  void attemptReady(OperatorCoordinator.SubtaskGateway gateway) {
+    Preconditions.checkState(this.coordinatorThreadFactory.isCurrentThreadCoordinatorThread());
+    this.subtaskGateways.registerSubtaskGateway(gateway);
+  }
+
+  void attemptFailed(int subtaskIndex, int attemptNumber) {
+    Preconditions.checkState(this.coordinatorThreadFactory.isCurrentThreadCoordinatorThread());
+    this.subtaskGateways.unregisterSubtaskGateway(subtaskIndex, attemptNumber);
+  }
+
+  void subtaskReset(int subtaskIndex) {
+    Preconditions.checkState(this.coordinatorThreadFactory.isCurrentThreadCoordinatorThread());
+    this.subtaskGateways.reset(subtaskIndex);
+  }
+
+  void failJob(Throwable cause) {
+    operatorCoordinatorContext.failJob(cause);
+  }
+
+  /**
+   * A helper method that delegates the callable to the coordinator thread if the current thread is
+   * not the coordinator thread, otherwise call the callable right away.
+   *
+   * @param callable the callable to delegate.
+   */
+  void callInCoordinatorThread(Callable<Void> callable, String errorMessage) {
+    // Ensure the task is done by the coordinator executor.
+    if (!coordinatorThreadFactory.isCurrentThreadCoordinatorThread()) {
+      try {
+        final Callable<Void> guardedCallable =
+            () -> {
+              try {
+                return callable.call();
+              } catch (Throwable t) {
+                LOG.error("Uncaught Exception in DataStatistics Coordinator Executor", t);
+                ExceptionUtils.rethrowException(t);
+                return null;
+              }
+            };
+
+        coordinatorExecutor.submit(guardedCallable).get();
+      } catch (InterruptedException | ExecutionException e) {
+        throw new FlinkRuntimeException(errorMessage, e);
+      }
+    } else {
+      try {
+        callable.call();
+      } catch (Throwable t) {
+        LOG.error("Uncaught Exception in DataStatistics coordinator executor", t);
+        throw new FlinkRuntimeException(errorMessage, t);
+      }
+    }
+  }
+
+  private static class SubtaskGateways {
+    private final Map<Integer, OperatorCoordinator.SubtaskGateway>[] gateways;
+
+    private SubtaskGateways(int parallelism) {
+      this.gateways = new Map[parallelism];
+
+      for (int i = 0; i < parallelism; ++i) {
+        this.gateways[i] = new HashMap<>();
+      }
+    }
+
+    private void registerSubtaskGateway(OperatorCoordinator.SubtaskGateway gateway) {
```

Comment:
wondering if we need debug logs for the subtask registrations/unregistrations/ready methods?